### PR TITLE
8340732: RISC-V: Refactor crc32 scalar version

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you help to review this patch?
As discussed in https://github.com/openjdk/jdk/pull/20910#discussion_r1755150447, it's helpful to refactor the existing scalar version of crc32 intrinsic.
Several refactoring are done in this pr,
1. Simplify the `len` usage, now it only decreases (i.e. change in one direction)
2. Simplify the code paths
3. Remove one instruction in `L_by4_loop`
4. Remove unnecessary code
5. Other misc

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340732](https://bugs.openjdk.org/browse/JDK-8340732): RISC-V: Refactor crc32 scalar version (**Enhancement** - P4)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21150/head:pull/21150` \
`$ git checkout pull/21150`

Update a local copy of the PR: \
`$ git checkout pull/21150` \
`$ git pull https://git.openjdk.org/jdk.git pull/21150/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21150`

View PR using the GUI difftool: \
`$ git pr show -t 21150`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21150.diff">https://git.openjdk.org/jdk/pull/21150.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21150#issuecomment-2370412041)

### CODE REVIEW

**Review by RealFYang — COMMENTED** [review]
Generally looks fine. I only have minor comments.

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 1]
```diff
@@ -6084,26 +6084,17 @@ static const int64_t right_3_bits = right_n_bits(3);
...
-    const Register tmp6 = x31; // t6
 
     BLOCK_COMMENT("Entry:");
     __ enter(); // required for proper stackwalking of RuntimeStub frame
 
-    __ kernel_crc32(crc, buf, len, table0, table1, table2,
-                    table3, tmp1, tmp2, tmp3, tmp4, tmp5, tmp6);
+    __ kernel_crc32(crc, buf, len,
+                    c_rarg3, c_rarg4, c_rarg5, c_rarg6, // tmp's for tables
+                    c_rarg7, t2, x28, x29, x30, x31);   // misc tmps
```
[the remark this answers is not in the document]
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:6097]: How do you think we add t3~t6 alias into assemler_riscv.hpp?
I saw several times people had the similar question, and in the code there are several places like this `Register size      = x29;   // t4`, i.e. try to use t3~t6, but needs to add a comment for that.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 2]
```diff
@@ -1604,56 +1603,49 @@ void MacroAssembler::kernel_crc32(Register crc, Register buf, Register len,
...
     update_word_crc32(crc, tmp1, tmp2, tmp4, tmp6, table0, table1, table2, table3, false);
-    subw(len, len, 4);
     addi(buf, buf, 4);
-    bge(len, zr, L_by4_loop);
-    addiw(len, len, 4);
-    ble(len, zr, L_exit);
+    blt(buf, loop_buf_end, L_by4_loop);
 
   bind(L_by1_loop);
+    blez(len, L_exit);
```
[the remark this answers is not in the document]
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:1621]: Yes, will do it and below ones.

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (RealFYang) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:6097]: Ah, I just realized we only have t0 - t2 defined in file assemler_riscv.hpp. I think its reasonable and helpful to add more ones like t3 - t6. These are temporary registers names / ABI Mnemonics specified by the RISC-V psABI spec.

[the remark this answers is not in the document]
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:6097]: OK, let me do it in another pr, as there're several other places like `Register size      = x29;   // t4`, I will change them together, and also include this place.
tracked by https://bugs.openjdk.org/browse/JDK-8340880

**Review by RealFYang — APPROVED** [review]
Updated change looks good to me. Any test performed? Thanks.

**Comment by Hamlin-Li** [comment]
> Updated change looks good to me. Any test performed? Thanks.

Thanks for reviewing!
I just did a test, there is a little difference from previous version, some perf gain, some perf loss when size <= 256, when size get bigger, there is no difference. Basically I think we are good.

**Comment by Hamlin-Li** [comment]
Thanks @RealFYang for your reviewing!

**Pull request closed without merging**